File: 2435-paths-in-matrix-whose-sum-is-divisible-by-k/2435-paths-in-matrix-whose-sum-is-divisible-by-k.py
class Solution:
    def numberOfPaths(self, grid: List[List[int]], k: int) -> int:
        m = len(grid)
        n = len(grid[0])
        MOD = 10**9 + 7
        dp = [[0] * k for _ in range(n)]
        
        for i in range(m):
            new_dp = [[0] * k for _ in range(n)]

            for j in range(n):
                vals = grid[i][j] % k
                if i == 0 and j == 0:
                    new_dp[j][vals] = 1
                    continue

                for r in range(k):
                    ways = 0
                    if i > 0:
                        ways += dp[j][r]
                    if j > 0:
                        ways += new_dp[j - 1][r]
                    if ways == 0:
                        continue
                    nr = (vals + r) % k    
                    new_dp[j][nr] = (new_dp[j][nr] + ways) % MOD

            dp, new_dp = new_dp, dp
        
        return dp[n-1][0]

File: 2435-paths-in-matrix-whose-sum-is-divisible-by-k/2435-paths-in-matrix-whose-sum-is-divisible-by-k.cpp
class Solution {
public:
    int numberOfPaths(vector<vector<int>>& grid, int k) {
        int m = grid.size();
        int n = grid[0].size();
        const int MOD = 1e9 + 7;
        
        vector<vector<int>>dp(n, vector<int>(k, 0));

        for (int i = 0; i < m ; i++) {
            vector<vector<int>>new_dp(n, vector<int>(k, 0));

            for (int j = 0; j < n; j++) {
                int val = grid[i][j] % k;

                if (i == 0 && j == 0) {
                    new_dp[j][val] = 1;
                    continue;
                }

                for (int r = 0; r < k; r++) {
                    long long ways = 0;
                    if (i > 0) {
                        ways += dp[j][r];
                    }
                    if (j > 0) {
                        ways += new_dp[j - 1][r];
                    }
                    if (ways == 0) {
                        continue;
                    }
                    int nr = (val + r) % k;
                    new_dp[j][nr] = (new_dp[j][nr] + ways) % MOD;
                }
                
            }
            swap(dp, new_dp);
        }
        return dp[n-1][0];
    }
};